JSON-querying API documentation refactor (#14589)
apache/druid: docs/api-reference/json-querying-api.md

### URL

<code class="postAPI">POST</code> <code>/druid/v2/</code>

### Query parameters

* `pretty` (optional)
  * Druid returns the response in a pretty-printed format using indentation and line breaks.

### Responses

<!--DOCUSAURUS_CODE_TABS-->

<!--200 SUCCESS-->

*Successfully submitted query* 

<!--400 BAD REQUEST-->

*Error thrown due to bad query. Returns a JSON object detailing the error with the following format:*

```json
{
    "error": "A well-defined error code.",
    "errorMessage": "A message with additional details about the error.",
    "errorClass": "Class of exception that caused this error.",
    "host": "The host on which the error occurred."
}
```
For more information on possible error messages, see [query execution failures](../querying/querying.md#query-execution-failures).

<!--END_DOCUSAURUS_CODE_TABS-->

---

> ### Sample request: `topN`

Reviewer: What if we made the request/response a single subheading since they're more closely related than the other stuff in this section?
![image](https://github.com/apache/druid/assets/7747997/b4edbce6-3a41-4830-80fc-8f48bb62d0b6)

So we could call this one:
```suggestion
### Example query: `topN`
```

And then either remove the response subheading or make it an H4 (####)

> The following example submits a JSON query of the `topN` type to retrieve a ranked list of users and their post views. 

Reviewer: ```suggestion
The following example shows a `topN` query. The query analyzes the `social_media` datasource to return the top five users from the `username` dimension with the highest number of views from the `views` metric.. 
```

<details>
  <summary>Click to show sample response</summary>

  ```json
[
    {
        "timestamp": "2023-07-03T18:49:54.848Z",
        "result": [
            {
                "views": 11591218026,
                "username": "gus"
            },
            {
                "views": 11578638578,
                "username": "miette"
            },
            {
                "views": 11561618880,
                "username": "leon"
            },
            {
                "views": 11552609824,
                "username": "mia"
            },
            {
                "views": 11551537517,
                "username": "milton"
            }
        ]
    }
]
  ```
</details>

### Sample request: `groupBy`

The following example submits a JSON query of the `groupBy` type to retrieve the `username` with the highest votes to posts ratio from the `social_media` datasource.

In this query:
* The `upvoteSum` aggregation calculates the sum of the `upvotes` for each user.
* The `postCount` aggregation calculates the sum of posts for each user.
* The `upvoteToPostRatio` is a post-aggregation of the `upvoteSum` and the `postCount`, divided to calculate the ratio.

> * The result is sorted based on the `upvoteToPostRatio` and in descending order.

Reviewer: ```suggestion
* The result is sorted based on the `upvoteToPostRatio` in descending order.
```

  ~ under the License.
  -->

This topic describes the API endpoints to submit JSON-based [native queries](../querying/querying.md) to Apache Druid.

In this topic, `http://SERVICE_IP:SERVICE_PORT` is a placeholder for the server address of deployment and the service port. For example, on the quickstart configuration, replace `http://ROUTER_IP:ROUTER_PORT` with `http://localhost:8888`.

## Submit a query

> Submits a JSON-based native query with the request body. 

Reviewer: ```suggestion
Submits a JSON-based native query. The body of the request is the native query itself.
```

Druid supports different types of queries for different use cases. All queries require the following properties:

> * `queryType`: A string representing type of query. Druid supports the following native query types: `timeseries`, `topN`, `groupBy`, `timeBoundaries`, `segmentMetadata`, `datasourceMetadata`, `scan`, and `search`. 

Reviewer: ```suggestion
* `queryType`: A string representing the type of query. Druid supports the following native query types: `timeseries`, `topN`, `groupBy`, `timeBoundaries`, `segmentMetadata`, `datasourceMetadata`, `scan`, and `search`. 
```

  "limitSpec": {
    "type": "default",
    "limit": 1,
    "columns": [
      { "dimension": "upvoteToPostRatio", "direction": "descending" }
    ]
  }
}'
```

<!--HTTP-->

```HTTP
POST /druid/v2?pretty=null HTTP/1.1
Host: http://ROUTER_IP:ROUTER_PORT
Content-Type: application/json
Content-Length: 817

{
  "queryType": "groupBy",
  "dataSource": "social_media",
  "dimensions": ["username"],
  "granularity": "all",
  "aggregations": [
    { "type": "doubleSum", "name": "upvoteSum", "fieldName": "upvotes" },
    { "type": "count", "name": "postCount", "fieldName": "post_title" }
  ],
  "postAggregations": [
    {
      "type": "arithmetic",
      "name": "upvoteToPostRatio",
      "fn": "/",
      "fields": [
        { "type": "fieldAccess", "name": "upvoteSum", "fieldName": "upvoteSum" },
        { "type": "fieldAccess", "name": "postCount", "fieldName": "postCount" }
      ]
    }
  ],
  "intervals": ["2022-01-01T00:00:00.000/2024-01-01T00:00:00.000"],
  "limitSpec": {
    "type": "default",
    "limit": 1,
    "columns": [
      { "dimension": "upvoteToPostRatio", "direction": "descending" }
    ]
  }
}
```

<!--END_DOCUSAURUS_CODE_TABS-->

> ### Sample response: `groupBy` query

Reviewer 2: ```suggestion
### Sample response: `groupBy`
```

#### Query parameters
* `pretty` (optional)
  * The response is returned in a pretty-printed format using indentation and line breaks.

#### Responses

<!--DOCUSAURUS_CODE_TABS-->

<!--200 SUCCESS-->
<br/>
*Successfully submitted query* 
<!--400 BAD REQUEST-->
<br/>
*Error thrown due to bad query. Returns a JSON object detailing the error with the following format:*

```json
{
    "error": "A well-defined error code.",
    "errorMessage": "A message with additional details about the error.",
    "errorClass": "Class of exception that caused this error.",
    "host": "The host on which the error occurred."
}
```
For more information on possible error messages, see [query execution failures](../querying/querying.md#query-execution-failures).
<!--END_DOCUSAURUS_CODE_TABS-->

---

#### Sample request

The following example submits a JSON query of the `topN` type to retrieve a ranked list of users and their post views. 

<!--DOCUSAURUS_CODE_TABS-->

<!--cURL-->
```shell

> curl "http://<ROUTER_IP>:<ROUTER_PORT>/druid/v2?pretty=null" \

Reviewer 2: ```suggestion
curl "http://ROUTER_IP:ROUTER_PORT/druid/v2?pretty=null" \
```

Reviewer 2: Where is the POST part of the curl request? Shouldn't it be
`curl -X POST`

--header "Content-Type: application/json" \
--data "{
  \"queryType\": \"topN\",
  \"dataSource\": \"social_media\",
  \"dimension\": \"username\",
  \"threshold\": 5,
  \"metric\": \"views\",
  \"granularity\": \"all\",

  \"aggregations\": [
    {
      \"type\": \"longSum\",
      \"name\": \"views\",
      \"fieldName\": \"views\"
    }
  ],
  \"intervals\": [
    \"2022-01-01T00:00:00.000/2024-01-01T00:00:00.000\"
  ]
}"
```
<!--HTTP-->
```HTTP
POST /druid/v2?pretty=null HTTP/1.1

> Host: http://<ROUTER_IP>:<ROUTER_PORT>

Reviewer 2: ```suggestion
Host: http://ROUTER_IP:ROUTER_PORT
```

Content-Type: application/json
Content-Length: 336

{
  "queryType": "topN",
  "dataSource": "social_media",
  "dimension": "username",
  "threshold": 5,
  "metric": "views",
  "granularity": "all",

  "aggregations": [
    {
      "type": "longSum",
      "name": "views",
      "fieldName": "views"
    }
  ],
  "intervals": [
    "2022-01-01T00:00:00.000/2024-01-01T00:00:00.000"
  ]
}
```
<!--END_DOCUSAURUS_CODE_TABS-->

The following example submits a JSON query of the `groupBy` type to retrieve the `username` with the highest upvotes to posts ratio from the `social_media` datasource.

In this query:
* The `upvoteSum` aggregation calculates the sum of the `upvotes` for each user.
* The `postCount` aggregation calculates the sum of posts for each user.
* The `upvoteToPostRatio` is a post-aggregation of the the `upvoteSum` and the `postCount`, divided to calculate the ratio.
* The result is sorted based on the `upvoteToPostRatio` and in descending order.

<!--DOCUSAURUS_CODE_TABS-->

<!--cURL-->
```shell

> curl "http://<ROUTER_IP>:<ROUTER_PORT>/druid/v2?pretty=null" \

Reviewer 2: ```suggestion
curl "http://ROUTER_IP:ROUTER_PORT/druid/v2?pretty=null" \
```

--header "Content-Type: application/json" \
--data "{
  \"queryType\": \"groupBy\",
  \"dataSource\": \"social_media\",
  \"dimensions\": [\"username\"],
  \"granularity\": \"all\",
  \"aggregations\": [
    { \"type\": \"doubleSum\", \"name\": \"upvoteSum\", \"fieldName\": \"upvotes\" },
    { \"type\": \"count\", \"name\": \"postCount\", \"fieldName\": \"post_title\" }
  ],
  \"postAggregations\": [
    {
      \"type\": \"arithmetic\",
      \"name\": \"upvoteToPostRatio\",
      \"fn\": \"/\",
      \"fields\": [
        { \"type\": \"fieldAccess\", \"name\": \"upvoteSum\", \"fieldName\": \"upvoteSum\" },
        { \"type\": \"fieldAccess\", \"name\": \"postCount\", \"fieldName\": \"postCount\" }
      ]
    }
  ],
  \"intervals\": [\"2022-01-01T00:00:00.000/2024-01-01T00:00:00.000\"],
  \"limitSpec\": {
    \"type\": \"default\",
    \"limit\": 1,
    \"columns\": [
      { \"dimension\": \"upvoteToPostRatio\", \"direction\": \"descending\" }
    ]
  }
}"
```
<!--HTTP-->
```HTTP
POST /druid/v2?pretty=null HTTP/1.1

> Host: http://<ROUTER_IP>:<ROUTER_PORT>

Reviewer 2: ```suggestion
Host: http://ROUTER_IP:ROUTER_PORT
```

                "username": "gus"
            },
            {
                "views": 11578638578,
                "username": "miette"
            },
            {
                "views": 11561618880,
                "username": "leon"
            },
            {
                "views": 11552609824,
                "username": "mia"
            },
            {
                "views": 11551537517,
                "username": "milton"
            }
        ]
    }
]
  ```

  The following is a sample response of the second query, retrieving the user with the highest upvotes to posts ratio.

  ```json
[
    {
        "version": "v1",
        "timestamp": "2022-01-01T00:00:00.000Z",
        "event": {
            "upvoteSum": 8.0419541E7,
            "upvoteToPostRatio": 69.53014661762697,
            "postCount": 1156614,
            "username": "miette"
        }
    }
]
  ```
</details>

### Get segment information for query

#### URL
<code class="postAPI">POST</code> `/druid/v2/candidates/`

Retrieves an array that contains objects with segment information, including the server locations associated with the provided query. 

#### Query parameters
* `pretty` (optional)

>   * The response is returned in a pretty-printed format using indentation and line breaks.

Reviewer 2: ```suggestion
  * Druid returns the response in a pretty-printed format using indentation and line breaks.
```

#### Responses

<!--DOCUSAURUS_CODE_TABS-->

<!--200 SUCCESS-->
<br/>
*Successfully retrieved segment information* 
<!--400 BAD REQUEST-->
<br/>
*Error thrown due to bad query. Returns a JSON object detailing the error with the following format:*

```json
{
    "error": "A well-defined error code.",
    "errorMessage": "A message with additional details about the error.",
    "errorClass": "Class of exception that caused this error.",
    "host": "The host on which the error occurred."
}
```
For more information on possible error messages, see [query execution failures](../querying/querying.md#query-execution-failures).
<!--END_DOCUSAURUS_CODE_TABS-->

---

#### Sample request

<!--DOCUSAURUS_CODE_TABS-->

<!--cURL-->
```shell

> curl "http://<ROUTER_IP>:<ROUTER_PORT>/druid/v2/candidates" \

Reviewer 2: Where is the POST part of the curl request? Shouldn't it be
`curl -X POST`

--header "Content-Type: application/json" \
--data "{
  \"queryType\": \"topN\",
  \"dataSource\": \"social_media\",
  \"dimension\": \"username\",
  \"threshold\": 5,
  \"metric\": \"views\",
  \"granularity\": \"all\",

  \"aggregations\": [
    {
      \"type\": \"longSum\",
      \"name\": \"views\",
      \"fieldName\": \"views\"
    }
  ],
  \"intervals\": [
    \"2020-01-01T00:00:00.000/2024-01-01T00:00:00.000\"
  ]
}"
```
<!--HTTP-->
```HTTP
POST /druid/v2/candidates HTTP/1.1

> Host: http://<ROUTER_IP>:<ROUTER_PORT>

Reviewer 2: ```suggestion
Host: http://ROUTER_IP:ROUTER_PORT
```
